Modules/posixmodule.c

```diff
@@ -3242,6 +3267,28 @@ os_lchown_impl(PyObject *module, path_t *path, uid_t uid, gid_t gid)
 #endif /* HAVE_LCHOWN */
 
 
+#ifdef MS_WINDOWS
+static PyObject *
+win32_wgetcwd(void)
+{
+    wchar_t wbuf[MAXPATHLEN], *wbuf2;
+    PyObject *resobj;
+
+    Py_BEGIN_ALLOW_THREADS
+    wbuf2 = win32_wgetcwd_buf(wbuf, Py_ARRAY_LENGTH(wbuf));
+    Py_END_ALLOW_THREADS
+
+    if (!wbuf2) {
+        return PyErr_SetFromWindowsErr(0);
+    }
+    resobj = PyUnicode_FromWideChar(wbuf2, wcslen(wbuf2));
+    if (wbuf2 != wbuf)
+        PyMem_RawFree(wbuf2);
+    return resobj;
+}
```

Comment:
Here's the suggested change to make `win32_wgetcwd()` a function that doesn't hold the GIL and always allocates the buffer.

```suggestion
static wchar_t *
win32_wgetcwd(LPDWORD plength)
{
    wchar_t *wbuf = NULL;
    DWORD size = 0;

    while (1) {
        wchar_t *temp;
        DWORD result = GetCurrentDirectoryW(size, wbuf);
        if (!result) {
            goto fail;
        }

        /* L'\0' is not counted in result on success. */
        if (result < size) {
            *plength = result;
            break;
        }

        temp = PyMem_RawRealloc(wbuf, result * sizeof(wchar_t));
        if (!temp) {
            SetLastError(ERROR_OUTOFMEMORY);
            goto fail;
        }
        wbuf = temp;
        size = result;
    }

    return wbuf;

fail:
    if (wbuf) {
        PyMem_RawFree(wbuf);
    }
    return NULL;
}
```

Reply:
I've updated the suggestion to implement `static wchar_t *win32_wgetcwd(const wchar_t *buf, const size_t size)` more faithfully to POSIX `getcwd()`. Like glibc, passing `buf` as `NULL` supports allocating the buffer. The caller is responsible for freeing an allocated buffer.

Reply:
I agree with the idea of inlining `win32_wgetcwd()` back into `posix_getcwd()`. In 4 years since my original PR handling of `use_bytes` was changed, and now it seems to make sense to have conversion from the native path type to Python's unicode/bytes in the same place for both Windows and Unix.

However, I can't see the value of introducing a pretty generic POSIX-like `win32_wgetcwd()` as you suggest. We need to get the current directory in just two places, and I don't see why our needs would change in the future.

I've force-pushed a new commit, which is mostly just a rebase of the original PR + inlining of `win32_wgetcwd()` and renaming of `win32_wgetcwd_buf()` to `win32_wgetcwd()`. ISTM that  `win32_wgetcwd()` is pretty simple. Having the caller to compare the returned buffer to the passed one is surely a wrinkle, but I don't think it matters, and overall I don't see why using your generic `win32_wgetcwd()` would be much better. But if there is some rationale that I'm missing, I'll be happy to revisit this.

Reply:
> But if there is some rationale that I'm missing, I'll be happy to revisit this.

There's no deep rationale. I just figured I may as well copy the Linux behavior. POSIX [`getcwd()`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/getcwd.html) only says that the behavior is "unspecified" when `buf` is `NULL`. Your code implements part of the extended behavior, given `buf` is `NULL` and `buf_size` is 0. However, `GetCurrentDirectoryW()` assumes that `nBufferLength` accurately represents the size of `lpBuffer`. If this size is big enough to hold the path of the current working directory, but `lpBuffer` is `NULL`, `GetCurrentDirectoryW()` will try to write to a `NULL` pointer and crash on an access violation. This can't happen in how we're currently using the function, but I felt it wouldn't hurt to round out the implementation. The Linux behavior is to allocate a new buffer in this case. Another option is to simply fail if `buf_size` is greater than 0 when `buf` is `NULL`.

Reply:
I see. I've intended `win32_getcwd()` purely as a helper for the two cases we have, not as a generic function (and one of those cases [shouldn't probably exist at all](https://github.com/python/cpython/pull/5802#issuecomment-1221621415)). I generally don't like to abstract over a small number of use cases. It works for `buf` is `NULL` and `buf_size` is 0 purely by accident. I can add an `assert()` or a failure case for `buf == NULL` if you like. 

I see how it could be tempting to add a more flexible version, but I think that would be justified if we needed to implemented `Py_getcwd()`  for `Python/fileutils.c`, but not here.